evalml/pipelines/components/transformers/samplers/oversamplers.py

```diff
@@ -0,0 +1,90 @@
+from evalml.pipelines.components.transformers.samplers.base_sampler import (
+    BaseOverSampler
+)
+
+
+class SMOTESampler(BaseOverSampler):
+    """SMOTE Oversampler component. Works on numerical datasets only. This component is only run during training and not during predict."""
+    name = "SMOTE Oversampler"
+    hyperparameter_ranges = {}
+
+    def __init__(self, sampling_ratio=0.25, sampling_ratio_dict=None, k_neighbors=5, n_jobs=-1, random_seed=0, **kwargs):
+        """Initialize the SMOTE Oversampler component.
+
+        Arguments:
+            sampling_ratio (float): This is the goal ratio of the minority to majority class, with range (0, 1]. A value of 0.25 means we want a 1:4 ratio
+                of the minority to majority class after oversampling. If the targets are multiclass, will create a dictionary using this ratio. Defaults to 0.25.
+            sampling_ratio_dict (dict): Dictionary which has keys corresponding to each class, and the values are the number of samples we want to oversample to for each class key.
+                If this value is provided, it will be used. Otherwise, we opt to use sampling_ratio. Defaults to None.
+            k_neighbors (int): The number of nearest neighbors to used to construct synthetic samples. Defaults to 5.
+            n_jobs (int): The number of CPU cores to use. Defaults to -1.
+        """
+        parameters = {"sampling_ratio": sampling_ratio,
+                      "sampling_ratio_dict": sampling_ratio_dict,
+                      "k_neighbors": k_neighbors,
+                      "n_jobs": n_jobs}
+        parameters.update(kwargs)
+        super().__init__("SMOTE",
+                         parameters=parameters,
+                         component_obj=None,
+                         random_seed=random_seed)
+
+
+class SMOTENCSampler(BaseOverSampler):
+    """SMOTENC Oversampler component. Uses SMOTENC to generate synthetic samples. Works on a mix of nomerical and categorical columns.
+       This component is only run during training and not during predict."""
+    name = "SMOTENC Oversampler"
+    hyperparameter_ranges = {}
+
+    def __init__(self, categorical_features=[], sampling_ratio=0.25, sampling_ratio_dict=None, k_neighbors=5, n_jobs=-1, random_seed=0, **kwargs):
```

Comment:
I think we should let woodwork determine which features are categorical features. Is the concern that the features created by the OHE are not marked as categorical?

In the meantime we should set `None` as the default value for `categorical_features`. Setting mutable default arguments can lead to bugs.

Reply:
@freddyaboulton I was going to have AutoMLSearch automatically detect the categorical features, but do you think it'd be better to have the component do that instead?

Reply:
I think that should be the desired end-goal. It reminds me of what we did for the `TextFeaturizer` in automl. At first, we had automl search detect the natural language columns and we had to pass that down all the way to the `IterativeAlgorithm` in order for the pipelines to use it. Eventually we let the component identify the natural language columns and I think the whole implementation is simpler because of it.

If for whatever reason we can't get to that desired end-goal right now, like `SmoteNC` needs to know that the features created by a OHE are also categorical (I'm just brainstorming here), I think it's fine to leave as-is and file issues to improve our components so that we can keep track of what we need to do in order to get rid of the `categorical_features` argument.